lazyServer.go

```diff
@@ -0,0 +1,33 @@
+package multistream
+
+import (
+	"io"
+	"sync"
+)
+
+type lazyServerConn struct {
+	waitForHandshake sync.Once
+	werr             error
+
+	con io.ReadWriteCloser
+}
+
+func (l *lazyServerConn) Write(b []byte) (int, error) {
+	l.waitForHandshake.Do(func() { panic("didn't initiate handshake") })
+	if l.werr != nil {
+		return 0, l.werr
+	}
+	return l.con.Write(b)
+}
+
+func (l *lazyServerConn) Read(b []byte) (int, error) {
+	// TODO: The tests require this for some reason. Not sure if it's correct...
```

Comment:
That *is* a bit weird.

Reply:
Ah, it could be that a net.Pipe acts strangely with zero length writes.

Reply:
You have a check that skips calling write if the length of the buffer is zero. The net.Pipe likely only unblocks a read once a write has been called, no matter the length involved.

Reply:
Yeah, that appears to be the case: https://golang.org/src/io/pipe.go?s=2550:2607#L47